### PR TITLE
Separate cooling cfl and reorder floors

### PR DESCRIPTION
**Pull request by pgrete**
I now implemented the changes we discussed in our last call, i.e.,
- the cooling cfl is now independent on the integrator (and just checks that the entries in the table are equally spaced)
- the pressure floor takes precedence in the `ConsToPrim` call (but the Tfloor is also called afterwards if both are present)
- in the cooling functions, only the Tfloor counts (not the pressure floor) 

In addition, I 
- documented cooling and floor parameters
- bumped Parthenon to include support for Ascent output
- removed the "mixed" integrator, which I put there some time ago but never really tested

If you're fine with those changes, feel free to merge and then merge the new `main` into the cluster branch.
If there are issues, I'll likely first be able to address them on/after 6 Mar.

### CODE REVIEW

**Review by BenWibking — APPROVED** [review]
Looks good to me.

File: src/hydro/srcterms/tabular_cooling.cpp [hunk 1]
```diff
@@ -460,7 +456,11 @@ void TabularCooling::SubcyclingFixedIntSrcTerm(MeshData<Real> *md, const Real dt
           sub_iter++;
         }
 
-        PARTHENON_REQUIRE(internal_e > internal_e_floor, "cooled below floor");
+        // If cooled below floor, reset to floor value.
+        // This could happen if the floor value is larger than the lower end of the
+        // cooling table or if they are close and the last subcycle in the cooling above
+        // the lower end pushed the temperature below the lower end (and the floor).
+        internal_e = (internal_e > internal_e_floor) ? internal_e : internal_e_floor;
```
Comment (forrestglines) [inline src/hydro/srcterms/tabular_cooling.cpp:463]: So this would allow cooling below the cooling table to occur if the cooling at the end of the table is fast. Should the temperature instead be fixed to the bottom of the cooling table?

Note that there is already a mechanism above to stop gas that is already below `internal_e_floor` from doing any cooling.

Maybe we should introduce a `cooling_internal_e_floor` and `global_internal_e_floor` below which, respectively, no cooling occurs and below which the internal energy is raised to match the global floor.

File: src/hydro/srcterms/tabular_cooling.cpp [hunk 2]
```diff
@@ -751,9 +634,11 @@ Real TabularCooling::EstimateTimeStep(MeshData<Real> *md) const {
...
 
         // Compute cooling time
-        // If de_dt is zero, using infinite cooling time
-        const Real cooling_time = (de_dt == 0 ? std::numeric_limits<Real>::infinity()
-                                              : fabs(internal_e / de_dt));
+        // If de_dt is zero (temperature is smaller than lower end of cooling table) or
+        // current temp is below floor, use infinite cooling time
+        const Real cooling_time = ((de_dt == 0) || (internal_e < internal_e_floor))
+                                      ? std::numeric_limits<Real>::infinity()
+                                      : fabs(internal_e / de_dt);
```
Comment (forrestglines) [inline src/hydro/srcterms/tabular_cooling.cpp:641]: Very good catch. I haven't tested much with a cooling "CFL" when gas is below the cooling table floor.

**Review by forrestglines — COMMENTED** [review]
@pgrete If you think the current approach is correct, feel free to merge.

However, I think that `internal_e_floor` is being confounded between a temperature floor for the hydro solver and a temperature floor for cooling. I think we should explicitly separate the two within the cooling table -- have cooling disabled below the cooling floor table but do not raise the temperature of gas enter the cooling function to this cooling floor.

File: docs/input.md [hunk 3]
```diff
@@ -161,3 +176,29 @@ Parameter: `max_dt` (float)
...
+integrator = townsend              # Other possible options are `rk12` and `rk45` for error bound subcycling
+#max_iter = 100                    # Max number of iteration for subcycling. Unsued for Townsend integrator
+cfl = 0.1                          # Restrict global timestep to `cfl*e / dedt`, i.e., some fraction of change per cycle in the specific internal energy (i.e., temperature)
+d_log_temp_tol = 1e-8              # Tolerance in cooling table between subsequent entries. Both subcycling integrators and cfl restriction rely on a table lookup that assumes equally spaced (in log space) temperature values.
+#d_e_tol = 1e-8                    # Tolerance for the relative error in the change of internal energy for the error bound subcyling integrators (rk12 and rk45). Unused for Townsend integrator.
+```
+
+*Note* several special cases for handling the lower end of the cooling table/low temperatures:
+- cooling is turned off once the temperature reaches the lower end of the cooling table
+- the explicit, global temperature floor `<hydro/Tfloor>` takes precedence over the lower end of the cooling table
```
[the remark this answers is not in the document]
Comment (pgrete) [reply docs/input.md:202]: > How does this work? This line read ambiguously to me. In what context does the global temperature floor "take precedence?" Does this mean:

What I was trying to say is, that "we always reset to the global `Tfloor` (if it exists) already in the cooling function (and not first in `ConsToPrim` if the final temperature of the cooling kernel is below `TFloor`".

Given that this kind of not matches any of your readings, we should probably rephrase that sentence as it's very unclear.
Any suggestions?

File: src/hydro/srcterms/tabular_cooling.cpp (hunk 1 above)
Comment (pgrete) [reply src/hydro/srcterms/tabular_cooling.cpp:463]: As I mentioned above, my intention here was that if we end up below the "global" temperature floor we should already apply it here and not first in the ConsToPrim update.
In practice this will likely have no significant impact.

If you agree that this approach is what we want then I should probably also update the Townsend cooling as for that cooling I do not have this check at the end (I only check for the low end of the cooling table).

Regarding a split between `cooling_internal_e_floor` and `global_internal_e_floor`, I'd prefer to leave it at the single, global parameter approach and that the cooling floor is implicitly handled by the lower end of the cooling table.
What do you think?

**Comment by pgrete** [comment]
> @pgrete If you think the current approach is correct, feel free to merge.
> 
> However, I think that `internal_e_floor` is being confounded between a temperature floor for the hydro solver and a temperature floor for cooling. I think we should explicitly separate the two within the cooling table -- have cooling disabled below the cooling floor table but do not raise the temperature of gas enter the cooling function to this cooling floor.

I guess I'm coming from the point of view that "a floor is floor, so we're already in a dangerous regime if needed".
For example, our reconstruction routines currently don't obey floors (which I also think in theory they should).
So from that point of view, "fixing" cells more often is just contributing to the robustness/self-consistency with in given sim (knowing that hitting floors is already bad a priori).

Any other thoughts, @bwoshea @BenWibking

**Comment by BenWibking** [comment]
I missed the original design discussion for this, but in other codes, I've usually seen floors implemented as an operator-split step that changes the conserved variables.

I don't think I've seen it implemented in the conserved -> primitive step. I guess there's no reason why it couldn't be there, but it might yield very strange results if the conserved vars don't return to physical values within the next few timesteps. Can you elaborate on the reason for setting the floor in the conserved to primitive step?

**Comment by BenWibking** [comment]
> For example, our reconstruction routines currently don't obey floors (which I also think in theory they should).

Is there an option to reconstruct $(\rho, v, T)$ instead of $(\rho, v, P)$? Then as long as the reconstruction is monotonic, there shouldn't be an issue.

**Comment by pgrete** [comment]
> I missed the original design discussion for this, but in other codes, I've usually seen floors implemented as an operator-split step that changes the conserved variables.

IIRC we didn't really discuss this aspect, which, as far as my current understanding of the discussion goes is: Is it reasonable to enforce a global floor in more than one kernel during a cycle?
 
> I don't think I've seen it implemented in the conserved -> primitive step. I guess there's no reason why it couldn't be there, but it might yield very strange results if the conserved vars don't return to physical values within the next few timesteps. Can you elaborate on the reason for setting the floor in the conserved to primitive step?

That's probably Athena legacy then. I don't recall exactly how Enzo or Pluto (or other codes, say Quokka :) ) handle this.
One appealing reason to have it embedded (at least there) is that there's virtually no performance hit as you already have all the data required in cache anyway.

> > For example, our reconstruction routines currently don't obey floors (which I also think in theory they should).
> 
> Is there an option to reconstruct (ρ,v→,T) instead of (ρ,v→,P)? Then as long as the reconstruction is monotonic, there shouldn't be an issue.

We currently reconstruct without knowledge of the var at hand, i.e., each variable is simply reconstructed based on "math" but not on "physics".
In other words, the reconstruction routines can reconstruct whatever variable/input you give them.

**Comment by BenWibking** [comment]
> > Is there an option to reconstruct (ρ,v→,T) instead of (ρ,v→,P)? Then as long as the reconstruction is monotonic, there shouldn't be an issue.
> 
> We currently reconstruct without knowledge of the var at hand, i.e., each variable is simply reconstructed based on "math" but not on "physics". In other words, the reconstruction routines can reconstruct whatever variable/input you give them.

Ok, I see. I was thinking this could be accomplished by changing the cons2prim kernel and a few lines in the Riemann solver to convert back to pressure. Do other kernels (besides the Riemann solver) assume that the primitive vars are always $(\rho, \vec{v}, P)$?

File: src/hydro/srcterms/tabular_cooling.cpp (hunk 1 above)
Comment (forrestglines) [reply src/hydro/srcterms/tabular_cooling.cpp:463]: The distinction between the two floors is that the bottom of the cooling table represents something physical and the `Tfloor` is something numerical. As I understand the cooling tables I'm using, cooling should stop at the floor of the cooling table since the cooling time at that point is much longer than the timescale of the simulation. `Tfloor` might be lower than the floor of the cooling table and other processes might lead to lower temperatures, but cooling itself shouldn't drop the temperature below the floor of the cooling table as that would be unphysical.

I think this confusion arose because originally when I wrote my cooling integrator, the only temperature floor I considered was the floor of the cooling table. The temperature floor for the fluid solver incorrectly supplanted that floor within the cooling function. For the Enzo-like runs this hasn't mattered since the floor to the fluid solver was set to the same floor as the cooling table.

**Comment by forrestglines** [comment]
> > @pgrete If you think the current approach is correct, feel free to merge.
> > However, I think that `internal_e_floor` is being confounded between a temperature floor for the hydro solver and a temperature floor for cooling. I think we should explicitly separate the two within the cooling table -- have cooling disabled below the cooling floor table but do not raise the temperature of gas enter the cooling function to this cooling floor.
> 
> I guess I'm coming from the point of view that "a floor is floor, so we're already in a dangerous regime if needed". For example, our reconstruction routines currently don't obey floors (which I also think in theory they should). So from that point of view, "fixing" cells more often is just contributing to the robustness/self-consistency with in given sim (knowing that hitting floors is already bad a priori).

I'll reply to this comment but I think I put this in my other comment: the floor to the cooling table has a physical motivation rather than numerical since the cooling time at the bottom of the table is much longer than the simulation time. Thus the cooling function alone shouldn't push temperature below the bottom of the cooling table regardless of any other floors in the code.

File: src/hydro/srcterms/tabular_cooling.cpp (hunk 1 above)
Comment (BenWibking) [reply src/hydro/srcterms/tabular_cooling.cpp:463]: I don't think either are physical. The cooling tables go down to 10^4 K, right? Ly-a cooling peaks there, so the cooling timescale should be extremely short.

**Review by forrestglines — APPROVED** [review]
Assuming tests are still passing, go ahead and merge this

**Comment by forrestglines** [comment]
@pgrete I don't have a working clang on hand to do the formatting. Is a github auto-formatting working or could you push a formatting fix?

**Comment by pgrete** [comment]
@par-hermes format

**Comment by pgrete** [comment]
works here too ;)

**Comment by forrestglines** [comment]
> @par-hermes format

I should have just tried it!
